core/src/com/unciv/logic/automation/civilization/DiplomacyAutomation.kt

```diff
@@ -461,4 +464,70 @@ object DiplomacyAutomation {
             .any { trade -> trade.trade.ourOffers.any { offer -> offer.name == offerName }
                     || trade.trade.theirOffers.any { offer -> offer.name == offerName } }
     }
+
+    internal fun denounce(
+        civInfo: Civilization
+    ) {
+        /*
+        this is not an official formula - modify or replace if needed
+        
+        whether to denounce is determined by how rapidly opinion has declined, the current relationship level, and (to do) personality traits
+        
+        with the current formula, excluding decounceWillingness, the AI will denounce if opinion drops rapidly from:
+            135 to 50
+            65 to 0
+            0 to -50
+            -60 to -100
+        
+        what "rapidly" means can be adjusted with DiplomacyManager#smoothedOpinionEmaPeriod
+        */
+
+        fun requiredOpinionChange(
+            diplomacyManager: DiplomacyManager,
+            denounceWillingnessModifier: Float = 1f
+        ): Float =
+            -65 * denounceWillingnessModifier * 1.005f.pow(diplomacyManager.opinionOfOtherCiv())
+
+        fun debugPrint(diplomacy: DiplomacyManager) {
```

Comment:
Please use the existing logging and logging-filtering instead.

Try inserting a `Log.debug("...")` and once imported Ctrl-B read the source comments there. That function can accept all your format parameters directly. It can even accept lambdas to defer formula evaluation, those run only when debug logging is on for your class.